junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertThrows.java

```diff
@@ -42,4 +45,25 @@
 			String.format("Expected %s to be thrown, but nothing was thrown.", getCanonicalName(expectedType)));
 	}
 
+	@SuppressWarnings("unchecked")
+	static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable,
+			Supplier<String> messageSupplier) {
+		return assertThrows(expectedType, executable, nullSafeGet(messageSupplier));
```

Comment:
Message suppliers must be evaluated lazily.

Please consult the implementations of other assertions for inspiration.